File: Objective-C/Dark Tower/Classes/DarkTower.m
//
//  DarkTower.m
//  DarkTower
//
//  Created by Louis Revor on 2/18/10.
//  Copyright 2010 Louis Revor. All rights reserved.
//

#import "DarkTower.h"
#import "SoundManager.h"
#import "Random.h"
#import "Dark_TowerViewController.h"
#import "MainMenuViewController.h"
#import "BazaarViewController.h"
#import "SetupViewController.h"
#import "BattleViewController.h"
#import "TowerStatusController.h"

@implementation DarkTower

static DarkTower *sharedInstance = nil;

@synthesize window;
@synthesize darkTowerController;
@synthesize mainMenuController;
@synthesize bazaarController;
@synthesize setupController;
@synthesize battleController;
@synthesize towerStatusController;

- (void)applicationDidFinishLaunching:(UIApplication *)application {    
    
    // Override point for customization after app launch  
    [window addSubview:towerStatusController.view];
    [window addSubview:battleController.view];
    [window addSubview:setupController.view];
    [window addSubview:bazaarController.view];
    [window addSubview:mainMenuController.view];
    [window addSubview:darkTowerController.view];
    [window makeKeyAndVisible];
}


- (void)dealloc {
    [towerStatusController release];
    [battleController release];
    [setupController release];
    [bazaarController release];
    [mainMenuController release];
    [darkTowerController release];
    [window release];
    [super dealloc];
}

- (id) init {
	if (self = [super init]) {
		numTurns = 0;
		level = 1;
		activePlayer=0;
		dragonGold = 0;
		dragonWarriors = 0;
	}
	sharedInstance=self;
	return self;
}

+ (DarkTower *) sharedInstance {
	if (!sharedInstance) [[self alloc] init];
	return sharedInstance;
}

- (void)reset {
	[myPlayer init:1];
	[SoundManager playSound:@"battle"];
}

- (void)laySiege{
	[SoundManager playSound:@"end-turn"];
    [window bringSubviewToFront:setupController.view];
}

- (void) startGame:(NSInteger) players withDifficulty:(NSInteger) difficulty {
//	NSInteger i;
	numPlayers = players;
	numTurns = 0;
	level = difficulty;
//	for (i=0; i<numPlayers; i++){
//		[aPlayers addObject: [[Player alloc] init:i]];
//	}
//	myPlayer = [aPlayers objectAtIndex: activePlayer];
	myPlayer = [[Player alloc] init:1];
	[SoundManager playSound:@"intro" withSleep: 5];
    [window bringSubviewToFront:mainMenuController.view];
}

- (void)endTurn {
    [window bringSubviewToFront:mainMenuController.view];
	[myPlayer endTurn];
}

- (void)bazaar {
	[myPlayer startTurn];
	if (![self cursed]) {
		[SoundManager playSound:@"bazaar"];
		[bazaarController setPrices: myPlayer.gold];
		[window bringSubviewToFront:bazaarController.view];
	}
}

- (void)endBazaar {
    [window bringSubviewToFront:mainMenuController.view];
	[myPlayer endTurn];
}

- (void)bazaarClosed {
	[towerStatusController bazaarClosed];
    [window bringSubviewToFront:towerStatusController.view];
	[myPlayer endTurn];
}

- (void)endBazaar:(NSInteger) gold 
	 withWarriors:(NSInteger) warriors
	  withRations:(NSInteger) rations
		withBeast:(BOOL) beast
		withScout:(BOOL) scout
	   withHealer:(BOOL) healer {
    [myPlayer setRations: [myPlayer rations]+rations];
    [myPlayer setWarriors: ([myPlayer warriors]+warriors)];
	if (beast) [myPlayer setBeast: TRUE];
	if (healer) [myPlayer setHealer: TRUE];
	if (scout) [myPlayer setScout: TRUE];
	[myPlayer setGold: ([myPlayer gold]-gold)];
	[myPlayer reduceGold];
	[self endTurn];
}

- (void)clear {
	[SoundManager playSound:@"clear"];
}

- (void)frontier {
	[myPlayer startTurn];
	if (![self cursed]) {
		[SoundManager playSound:@"frontier" withSleep: 1];
		switch (myPlayer.territory) {
			case 1:
				[myPlayer setTerritory: 2];
				[myPlayer endTurn];
				break;
			case 2:
				if (myPlayer.bronzeKey) {
					[myPlayer setTerritory: 3];
				} else {
					[window bringSubviewToFront:towerStatusController.view];
					[towerStatusController keyMissing];
				}
				break;
			case 3:
				if (myPlayer.silverKey) {
					[myPlayer setTerritory: 4];
				} else {
					[window bringSubviewToFront:towerStatusController.view];
					[towerStatusController keyMissing];
				}
				break;
			case 4:
				[window bringSubviewToFront:towerStatusController.view];
				[towerStatusController keyMissing];
				break;
		}
	}
}

- (void)inventory {
}

- (void)move {
	NSInteger move;
	[myPlayer startTurn];
	if (![self cursed]) {
		move = [Random getRand];
		//move = 8;
		switch (move) {
			case 1:
			case 2:
			case 3:
			case 4:
			case 5:
				[SoundManager playSound:@"beep"]; // nothing
				break;
			case 6: // Dragon
				if (myPlayer.dragonSword) {
					[myPlayer setGold: (myPlayer.gold+dragonGold)];
					[myPlayer reduceGold];
					[myPlayer setWarriors: (myPlayer.warriors+dragonWarriors)];
					dragonGold=0;
					dragonWarriors=0;
				} else {
					[myPlayer setGold: (myPlayer.gold * 0.75)];
					[myPlayer reduceGold];
					[myPlayer setWarriors: (myPlayer.warriors * 0.75)];
					dragonGold=(myPlayer.gold * 0.75);
					dragonWarriors=(myPlayer.warriors * 0.75);
				}
				[window bringSubviewToFront:towerStatusController.view];
				[towerStatusController dragon: myPlayer.dragonSword withGold: myPlayer.gold withWarriors: myPlayer.warriors];
				[myPlayer setDragonSword:FALSE];
				break;
			case 7: // lost
				[window bringSubviewToFront:towerStatusController.view];
				[towerStatusController lost: myPlayer.scout];
				break;
			case 8: // plague
				[window bringSubviewToFront:towerStatusController.view];
				if (myPlayer.healer) {
					[myPlayer setWarriors: (myPlayer.warriors + 2)];
				} else {
					[myPlayer setWarriors: (myPlayer.warriors - 2)];
				}
				[towerStatusController plague: myPlayer.healer withWarriors: myPlayer.warriors];
				break;
			case 9:
			case 10: // battle
				[SoundManager playSound:@"battle"];
				[self battle];
				break;
		}
		[myPlayer endTurn];
	}
}

- (void)pegasus{
	if (![self cursed]) {
		[SoundManager playSound:@"pegasus" withSleep: 3];
	}
}

- (void)sanctuary {
	[myPlayer startTurn];
	if (![self cursed]) {
		[SoundManager playSound:@"sanctuary"];
		[myPlayer sanctuary];
		[myPlayer endTurn];
	}
}

- (void)tomb {
	NSInteger value;
	[myPlayer startTurn];
	if (![self cursed]) {
#ifdef TOWERDEBUG
		value = 3;
#else
		value = [Random getRand];
#endif
		switch (value) {
			case 1:
			case 2:
				[SoundManager playSound:@"tomb-nothing" withSleep:4];
				[myPlayer endTurn];
				break;
			case 3:
				[SoundManager playSound:@"tomb" withSleep:4];
				[self treasure];
				break;
			default:
				[SoundManager playSound:@"tomb-battle" withSleep:4];
				[self battle];
				break;
		}
	}
}
- (void)tower {
	[myPlayer startTurn];
	if (![self cursed]) {
		if ((myPlayer.bronzeKey) && (myPlayer.silverKey) && (myPlayer.goldKey)) {
			[SoundManager playSound:@"darktower"];
			[myPlayer endTurn];
		} else {
			[window bringSubviewToFront:towerStatusController.view];
			[towerStatusController keyMissing];
		}
	}
}

- (void)treasure {
	BOOL gotTreasure = FALSE;
	while (!gotTreasure) {
		switch ([Random getRand]) {
			case 1:
				//Nothing
				gotTreasure=TRUE;
				NSLog(@"Treasure: NOTHING");
				[self endTurn];
				break;
			case 2:
			case 3:
			case 4:
				//Gold
				gotTreasure=TRUE;
				NSLog(@"Treasure: Gold");
				[myPlayer setGold:(myPlayer.gold + [Random getRand] + 10)];
				[myPlayer reduceGold];
				[window bringSubviewToFront:towerStatusController.view];
				[towerStatusController gold: myPlayer.gold];
				break;
			case 5:
			case 6:
				//Sword
				if (!myPlayer.dragonSword) {
					[myPlayer setDragonSword:TRUE];
					gotTreasure=TRUE;
					NSLog(@"Treasure: Sword");
					[window bringSubviewToFront:towerStatusController.view];
					[towerStatusController sword];
				}
				break;
			case 7:
				//Wizard
				if (numPlayers>1) {
					// Need to be able to select the player to curse and steal 1/4 of their
					// gold and warriors.
					// The cursed player action works
					//[myPlayer setCursed:TRUE];
					//gotTreasure=TRUE;
					//NSLog(@"Treasure: Wizard");
					//[window bringSubviewToFront:towerStatusController.view];
					//[towerStatusController wizard];
				}
				break;
			case 8:
				//Pegasus
				if (!myPlayer.pegasus) {
					[myPlayer setPegasus:TRUE];
					gotTreasure=TRUE;
					NSLog(@"Treasure: Pegasus");
					[window bringSubviewToFront:towerStatusController.view];
					[towerStatusController pegasus];
				}
				break;
			case 9:
			case 10:
				//Key
				switch (myPlayer.territory) {
					case 1:
						break;
					case 2:
						//Bronze Key
						if (!myPlayer.bronzeKey) {
							[myPlayer setBronzeKey:TRUE];
							gotTreasure=TRUE;
							NSLog(@"Treasure: Bronze Bey");
							[window bringSubviewToFront:towerStatusController.view];
							[towerStatusController bronzeKey];
						}
						break;
					case 3:
						//Silver Key
						if (!myPlayer.silverKey) {
							[myPlayer setSilverKey:TRUE];
							gotTreasure=TRUE;
							NSLog(@"Treasure: Silver Key");
							[window bringSubviewToFront:towerStatusController.view];
							[towerStatusController silverKey];
						}
						break;
					case 4:
						//Gold Key
						if (!myPlayer.goldKey) {
							[myPlayer setGoldKey:TRUE];
							gotTreasure=TRUE;
							NSLog(@"Treasure: Gold Key");
							[window bringSubviewToFront:towerStatusController.view];
							[towerStatusController goldKey];
						}
						break;
				}
				break;
		}
	}
}

- (void) battle {
    [window bringSubviewToFront:battleController.view];
	[battleController startBattle: myPlayer.warriors withLevel: level];
}

- (void) endBattle:(NSInteger) remainingWarriors withRetreat: (BOOL) retreat{
	[myPlayer setWarriors: remainingWarriors];
	if ((remainingWarriors>0) && (!retreat)) {
		[self treasure];
	} else {
		[self endTurn];
	}
}

- (BOOL)cursed {
	if (myPlayer.cursed) {
		[myPlayer setCursed:FALSE];
		[window bringSubviewToFront:towerStatusController.view];
		[towerStatusController cursed:myPlayer.warriors withGold:myPlayer.gold];
		return TRUE;
	} else {
		return FALSE;
	}
}

- (NSInteger)activeRations {
	return [myPlayer rations];
}

- (NSInteger)activeGold {
	return [myPlayer gold];
}

- (NSInteger)activeWarriors {
	return [myPlayer warriors];
}

@end

File: Swift/DarkTowerSwift/DarkTower/DarkTower.swift
//
//  DarkTower.swift
//  DarkTower
//
//  Created by Louis Revor on 2/28/15.
//  Copyright (c) 2015 Louis Revor. All rights reserved.
//

import Foundation
import UIKit

class DarkTower: NSObject, UIApplicationDelegate {    

    enum towerSegueState {
        case none, bazaar, status, battle, inventory, riddle
    }
    enum towerTurnState {
        case none, bazaar, tomb, sanctuary, move, frontier, inventory, darkTower, clear, pegasus
    }
    
    // Game Information
    //var myPlayer:Player! = nil
    var players:[Player!] = [nil, nil, nil, nil]
    var riddle:[NSInteger] = [0,0,0]
    var currentPlayer = 0
    
    //	NSArray aPlayers;
    var numPlayers:NSInteger=1
    var numTurns:NSInteger=0
    var level:NSInteger=1
    var activePlayer:NSInteger=0
    var dragonGold:NSInteger=10
    var dragonWarriors:NSInteger=5
    var random = Random()
    var goldAwarded=0
    var warriorsAwarded=0
    var rationsAwarded=0

    var currentTurnState = towerTurnState.none
    var segueState  = towerSegueState.none
    var statusState = TowerStatusController.state.none
    var battleState = BattleViewController.state.none

    func applicationDidFinishLaunching(application:UIApplication?) {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
    }
    
    class func sharedInstance() -> DarkTower! {
        // NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
       struct Static {
            static var instance: DarkTower? = nil
            static var onceToken: dispatch_once_t = 0
        }
        
        dispatch_once(&Static.onceToken) {
            Static.instance = self()
        }
        
        return Static.instance!
    }
    
    required override init() {
        super.init()
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
    }

    func setRiddle() {
        riddle[0]=0
        riddle[1]=0
        riddle[2]=0

        //set the bronze Key
        riddle[random.getRand(3)-1] = 1
        
        let silverKey = random.getRand(2)
        var empties=0
        for iter in 0..<3 {
            if (riddle[iter] == 0) {
                empties++
                if (empties == silverKey) { riddle[iter]=2 }
            }
        }

        for iter in 0..<3 {
            if (riddle[iter] == 0) { riddle[iter]=3 }
        }
        NSLog("[%@ %@] Riddle Sequence is: %d %d %d", reflect(self).summary, __FUNCTION__, riddle[0], riddle[1], riddle[2])
    }
    
    func reset() {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        numPlayers=1
        numTurns=0
        level=1
        activePlayer=0
        dragonGold=0
        dragonWarriors=0
        var iter:NSInteger=0
        for iter in 0..<4 {
            players[iter]=Player()
            players[iter].playerID=iter
        }
        setRiddle()
        SoundManager.playSound("battle")
    }
    
    func player() ->Player! {
       return(players[currentPlayer])
    }
    
    func startGame (players:NSInteger, difficulty:NSInteger) {
        numPlayers = players;
        numTurns = 0;
        level = difficulty;
        SoundManager.playSound("intro")
    }

    func startTurn(turnState:towerTurnState) {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        if (!dead()) {
            if (!cursed()) {
                currentTurnState = turnState
                if (currentTurnState != towerTurnState.pegasus) { player().startTurn() }
                runTurn()
            }
        }
    }
    
    func runTurn() {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        switch currentTurnState {
            case .bazaar:
                segueState  = towerSegueState.bazaar
            case .inventory:
                segueState  = towerSegueState.inventory
            case .tomb:
                segueState  = towerSegueState.battle
                battleState = BattleViewController.state.tomb
            case .sanctuary:
                if (player().warriors<5) {
                    player().warriors += random.getRand(5) + 7
                    warriorsAwarded = player().warriors
                }
                if (player().gold<8) {
                    player().gold += random.getRand(10) + 15
                    goldAwarded = player().gold
                }
                if (player().rations<6) {
                    player().rations += random.getRand(10) + 15
                    rationsAwarded = player().rations
                }
                segueState  = towerSegueState.status
                statusState = TowerStatusController.state.sanctuary
            case .move:
                var move:NSInteger=0
                move = random.getRand(10)
                switch (move) {
                    case 6: // Dragon
                        if (player().dragonSword) {
                            player().dragonSword=false
                            goldAwarded = dragonGold
                            warriorsAwarded = dragonWarriors
                            player().gold += goldAwarded
                            player().warriors += warriorsAwarded
                    
                            // reset back to original
                            dragonGold=10
                            dragonWarriors=5
                    
                            segueState  = towerSegueState.status
                            statusState = TowerStatusController.state.dragonKill
                        } else {
                            goldAwarded = player().gold / 4
                            warriorsAwarded = player().warriors / 4
                            player().gold -= goldAwarded
                            player().warriors -= warriorsAwarded
                            dragonGold += goldAwarded
                            dragonWarriors += warriorsAwarded
                        
                            segueState  = towerSegueState.status
                            statusState = TowerStatusController.state.dragon
                        }
                    case 7: // lost
                        segueState  = towerSegueState.status
                        statusState = TowerStatusController.state.lost
                    case 8: // plague
                        if (player().healer) {
                            player().warriors = player().warriors + 2
                            } else {
                                player().warriors = player().warriors - 2
                            }
                        segueState  = towerSegueState.status
                    statusState = TowerStatusController.state.plague
                    case 9,10: // battle
                        segueState  = towerSegueState.battle
                        battleState = BattleViewController.state.battle
                    default:
                        segueState  = towerSegueState.status
                        statusState = TowerStatusController.state.clear
                }
            case .frontier:
                if ((player().territory==1)&&(player().bronzeKey)) {
                    segueState  = towerSegueState.status
                    statusState = TowerStatusController.state.clear
                } else if ( ((player().territory==1)&&(!player().bronzeKey)) ||
                            ((player().territory==2)&&(player().bronzeKey)) ||
                            ((player().territory==3)&&(player().silverKey)) ||
                            ((player().territory==4)&&(player().goldKey)) ) {
                    player().territory+=1
                    if (player().territory==5) {player().territory=1}
                    segueState  = towerSegueState.status
                    statusState = TowerStatusController.state.frontier
                } else {
                    segueState  = towerSegueState.status
                    statusState = TowerStatusController.state.missingKey
                }
            case .darkTower:
                if ((player().territory != 1)||(!player().goldKey)) {
                    segueState  = towerSegueState.status
                    statusState = TowerStatusController.state.missingKey
                } else {
                    segueState = towerSegueState.riddle
                }
            case .clear:
                segueState  = towerSegueState.status
                statusState = TowerStatusController.state.clear
            case .pegasus:
                player().pegasus=false
                segueState  = towerSegueState.status
                statusState = TowerStatusController.state.pegasus
            default:
                NSLog("[%@ %@] Default - Should not get here", reflect(self).summary, __FUNCTION__)
        }
    }
    
    func endTurn() {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)

        if (currentTurnState != towerTurnState.pegasus) {
            player().endTurn()
            currentPlayer = (currentPlayer+1)%numPlayers
            SoundManager.playSound("end-turn")
        } else {
            SoundManager.playSound("beep")
        }
        
        // clear state
        currentTurnState = towerTurnState.none
        segueState  = towerSegueState.none
        statusState = TowerStatusController.state.none
        battleState = BattleViewController.state.none
        
        // clear awards
        goldAwarded=0
        warriorsAwarded=0
        rationsAwarded=0
    }
    
    func dead ()->CBool {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        if (player().warriors == 0) {
            segueState  = towerSegueState.status
            statusState = TowerStatusController.state.death
            return(true)
        } else {
            return(false)
        }
    }
    
    func cursed ()->CBool {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        if (player().cursed) {
            segueState  = towerSegueState.status
            statusState = TowerStatusController.state.cursed
            goldAwarded=player().gold
            warriorsAwarded=player().warriors
            player().cursed=false
            return(true)
        } else {
            return(false)
        }
    }
    
    func curse (playerID:NSInteger) {
        NSLog("[%@ %@] Curse: Player %02d is cursing Player %02d", reflect(self).summary, __FUNCTION__, player().playerID, playerID)
        goldAwarded = players[playerID].gold / 4
        warriorsAwarded = players[playerID].warriors / 4
        players[playerID].gold -= goldAwarded
        players[playerID].warriors -= warriorsAwarded
        players[playerID].cursed = true
        player().gold += goldAwarded
        player().warriors += warriorsAwarded
        goldAwarded = player().gold
        warriorsAwarded = player().warriors
        statusState = TowerStatusController.state.gold
        player().dump()
        players[playerID].dump()
    }

    func closeBazaar () {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        statusState = TowerStatusController.state.bazaarClosed
    }
    
    func winBattle () {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        var gotTreasure:CBool = false
        while (!gotTreasure) {
            switch (random.getRand()) {
            case 1:
                //Nothing
                gotTreasure=true
                NSLog("[%@ %@] Treasure: NOTHING", reflect(self).summary, __FUNCTION__)
                statusState = TowerStatusController.state.clear
            case 2,3,4:
                //Gold
                gotTreasure=true
                NSLog("[%@ %@] Treasure: Gold", reflect(self).summary, __FUNCTION__)
                player().gold = player().gold + random.getRand() + 10
                player().reduceGold()
                goldAwarded=player().gold
                statusState = TowerStatusController.state.gold
            case 5,6:
                //Sword
                if (!player().dragonSword) {
                    player().dragonSword=true
                    gotTreasure=true
                    NSLog("[%@ %@] Treasure: Sword", reflect(self).summary, __FUNCTION__)
                    statusState = TowerStatusController.state.sword
                }
                break;
            case 7:
                //Wizard
                if (numPlayers>1) {
                    gotTreasure=true
                    NSLog("[%@ %@] Treasure: Wizard!", reflect(self).summary, __FUNCTION__)
                    statusState = TowerStatusController.state.wizard
                }
            case 8:
                //Pegasus
                if (!player().pegasus) {
                    player().pegasus=true
                    gotTreasure=true
                    NSLog("[%@ %@] Treasure: Pegasus", reflect(self).summary, __FUNCTION__)
                    statusState = TowerStatusController.state.pegasus
                }
            case 9,10:
                //Key
                switch (DarkTower.sharedInstance().player().territory) {
                case 1:
                    NSLog("[%@ %@] Treasure: Key - not available in this territory", reflect(self).summary, __FUNCTION__)
                case 2:
                    //Bronze Key
                    if (!player().bronzeKey) {
                        player().bronzeKey=true
                        gotTreasure=true
                        NSLog("[%@ %@] Treasure: Bronze Key", reflect(self).summary, __FUNCTION__)
                        statusState = TowerStatusController.state.bronzeKey
                    }
                case 3:
                    //Silver Key
                    if (!player().silverKey) {
                        player().silverKey=true
                        gotTreasure=true
                        NSLog("[%@ %@] Treasure: Silver Key", reflect(self).summary, __FUNCTION__)
                        statusState = TowerStatusController.state.silverKey
                    }
                case 4:
                    //Gold Key
                    if (!player().goldKey) {
                        player().goldKey=true
                        gotTreasure=true
                        NSLog("[%@ %@] Treasure: Gold Key", reflect(self).summary, __FUNCTION__)
                        statusState = TowerStatusController.state.goldKey
                    }
                default:
                    NSLog("[%@ %@] Should not get here", reflect(self).summary, __FUNCTION__)
                }
            default:
                NSLog("[%@ %@] Should not get here", reflect(self).summary, __FUNCTION__)
                
            }
        }
    }
    
    func setBrigands (warriors:NSInteger)->NSInteger {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        var brigands:NSInteger=0
        
        switch (level) {
        case 1:
            brigands = warriors - 3 + random.getRand(100)*7/100;
        case 2:
            brigands = warriors + random.getRand(100)*6/100;
        case 3:
            brigands = warriors + 5 + random.getRand(100)*11/100;
        case 4:
            brigands = 16;
        default:
            NSLog("[%@ %@] Shouldn't get here", reflect(self).summary, __FUNCTION__)
        }
        NSLog("[%@ %@] Battle!  Warriors: %d, Brigands %d", reflect(self).summary, __FUNCTION__, warriors, brigands)
        return(brigands)
    }
    
    func setWarriorWin (warriors:NSInteger, brigands:NSInteger)->CBool {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        var probability:NSInteger=0
        
        if (warriors>brigands) {
            probability = 75 - (brigands*100)/(4*warriors);
        } else {
            probability = 25 + (warriors*100)/(4*brigands);
        }
        
        NSLog("[%@ %@] Probability of Win: %d", reflect(self).summary, __FUNCTION__, probability)
        if (random.getRand(100) < probability) {
            return(true)
        } else {
            return(false)
        }
    }
    
    func wrongRiddle () {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        statusState = TowerStatusController.state.missingKey
    }
    
    func correctRiddle () {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        battleState = BattleViewController.state.darkTower
    }
    
    func winDarkTower () {
        NSLog("[%@ %@]", reflect(self).summary, __FUNCTION__)
        statusState = TowerStatusController.state.victory
    }
}